### DISCUSSION TITLE
Focus in wrong window after switching workspaces, very confusing

### DISCUSSION THREAD

**Discussion by gsstark** [i3/i3#4798] [Q&A]
When I switch workspaces it seems quite random which window gets the focus. It seems more often than not it's not the window the mouse is in which feels very very wrong. What's worse, I've gotten into the habit of reaching for the mouse after switching workspaces to  verify that it's in the window I want and sometimes even after moving the mouse I then type and the focus goes to some other window. 

In an ideal world the focus would always be where the mouse is but I'm not sure how to maintain that when I switch windows within a workspace. But at the very least to avoid massive confusion I need a mode that when I switch workspaces at least start with the focus in the window the mouse is currently in.

**Comment by Airblader** [comment]
There's a few variables here, but if you switch workspaces on the same output without moving the mouse, the focused window will be whatever window was last focused when you were on that workspace before.

> In an ideal world the focus would always be where the mouse

I believe you would quickly find that much more annoying than you anticipate, but to each their own. :-) There's no configuration in i3 to achieve this, but you can certainly write a script using the IPC: listen for events when a workspace is switched, query the window at the cursor position and focus it.

**Reply by gsstark** [reply]
> There's a few variables here, but if you switch workspaces on the same output without moving the mouse, the focused window will be whatever window was last focused when you were on that workspace before.

I guess I could imagine a case where you're quickly switching back and forth between workspaces where that's intuitive. But my experience is that switching workspaces is a context switch and the computer is apparently much better at remembering where my focus was 20min ago than I am. Which is not surprising because I'm not expecting to get a pop quiz on that and haven't even tried to remember it.

I think instead of using IPC (which I fear would have race conditions and latency and be unnecessarily complex) it would make more sense to change my key workspace changing keybindings to have a chain of commands, one to switch workspace and then one to set the focus to the follow the mouse. But the focus command doesn't have any option to select the window with the mouse in it. 

So my use case would be satisfied by having window filter to select things like "the window under the mouse". That seems like it would be a generally useful feature too for other purposes.

**Reply by Airblader** [reply]
The IPC is good and fast enough for i3 to use it internally itself, so there's no reason to believe it has too much latency for you to use.

Since no command criterion for the cursor position exists, you need to first query the cursor position and find the window there and then pass this as the criterion to i3. I believe xdotool can do all the heavy lifting for you here with [something like](https://unix.stackexchange.com/a/70749)

```
xdotool getmouselocation --shell | grep WINDOW | awk -F "=" '{print $2}'
```

**Comment by gsstark** [comment]
Fwiw I just stumbled across yet another case where the focus ends up in the "wrong" window. When I close a tiling window the remaining tiled windows rejigger themselves and the focus seems to end up in a random window which isn't always the same one the mouse is in.

**Reply by Airblader** [reply]
I don't know why you expect it to be the window where the mouse is in; that's just not how i3 works, but it certainly isn't random. Focus goes to the last focused window within the parent of the closed window, recursing the tree upwards if necessary.

**Comment by gsstark** [comment]
So I have three solutions. They all seem workable but actually the simplest one seems to be working the best.

Your proposed solution using getmouselocation and setting focus to match:
`bindsym $mod+1 workspace 1; exec --no-startup-id i3-msg "[id=$(xdotool getmouselocation --shell | awk -F = '/^WINDOW/ {print $2}')] focus"`

Alternately warping the mouse on every workspace change to match the focus:
`bindsym $mod+2 workspace 2; exec --no-startup-id xdotool getactivewindow mousemove --window %1 0 0`

But in the process of looking for those I noticed this line in the xdotool manual:
```
            # Emulate focus-follows-mouse
            xdotool search . behave %@ mouse-enter windowfocus
```

Which actually seems to be working the best of the three. And it's the simplest.

**Reply by Airblader** [reply]
Note that i3 already has focus follows mouse (unless you disabled it), but we temporarily disable it for some situations. If you have this active it might be worth disabling it in i3.
